File: kreuzberg/packages/php/src/Types/CodeDocSection.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Section within a docstring.
 *
 * @property-read string $kind Section kind (e.g. 'param', 'returns', 'description')
 * @property-read string|null $name Section name (e.g. parameter name)
 * @property-read string $content Section content text
 */
readonly class CodeDocSection
{
    public function __construct(
        public string $kind,
        public ?string $name,
        public string $content,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $kind */
        $kind = $data['kind'] ?? '';

        /** @var string|null $name */
        $name = $data['name'] ?? null;

        /** @var string $content */
        $content = $data['content'] ?? '';

        return new self(
            kind: $kind,
            name: $name,
            content: $content,
        );
    }
}

File: kreuzberg/packages/elixir/lib/kreuzberg/code_doc_section.ex
defmodule Kreuzberg.CodeDocSection do
  @moduledoc """
  Section within a docstring.

  ## Fields

    * `:kind` - Section kind (e.g. "param", "returns", "description")
    * `:name` - Section name (e.g. parameter name)
    * `:content` - Section content text
  """

  @type t :: %__MODULE__{
          kind: String.t(),
          name: String.t() | nil,
          content: String.t()
        }

  defstruct kind: "",
            name: nil,
            content: ""

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    %__MODULE__{
      kind: Map.get(data, "kind", Map.get(data, :kind, "")),
      name: Map.get(data, "name", Map.get(data, :name)),
      content: Map.get(data, "content", Map.get(data, :content, ""))
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = section) do
    %{
      "kind" => section.kind,
      "name" => section.name,
      "content" => section.content
    }
  end
end
